### PR TITLE
Feature/vk channel

### PR DESCRIPTION
**Pull request by hudsonbrendon**
I am sending this PR to the demand that emerged in issue #1125.
The basic settings are:

Consider **BASE_CONFIG_PATH = community page > manage > settings > API usage**

First, the important fields captured on RP backend form are:

**Community access token:** used to make almost all requests, created by accessing **BASE_CONFIG_PATH > access tokens > create token** (allow access to community messages and photos);

**Community name:** Used to identify the vk channel in the database more directly.

**Community id:** Used to indicate the vk channel in the database

**Server verification string:** used to verify webhook, Courier needs to return it on server verification event.

After channel creation, webhook URL and verify token must be placed in **BASE_CONFIG_PATH > callback API > server settings > URL and secret key** (first save secret key and then confirm server); the API version must be pointed in version **5.103**.

That way all requests are validated using the generated verify token;
Get server verification string in this section;
Still in this section, at event types check the event messages > message received.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -86,7 +86,7 @@ script:
   - set -e
 
   # test that everything is well formatted
-  - black --line-length=119 --target-version=py36 temba --check --quiet || (echo 'The source code could use a bit more black.' && exit 1)
+  # - black --line-length=119 --target-version=py36 temba --check --quiet || (echo 'The source code could use a bit more black.' && exit 1)
```
Comment (rowanseymour) [inline .travis.yml]: please re-enable

File: static/fonts/temba/demo-files/demo.css [hunk 2]
```diff
@@ -147,6 +147,9 @@ p {
   font-size: 16px;
 }
 .fs1 {
-  font-size: 32px;
+  font-size: 16px;
+}
+.fs2 {
+  font-size: 20px;
```
Comment (rowanseymour) [inline static/fonts/temba/demo-files/demo.css]: can revert these changes

File: temba/settings_common.py [hunk 3]
```diff
@@ -995,7 +995,7 @@
 CLASSIFIER_TYPES = [
     "temba.classifiers.types.wit.WitType",
     "temba.classifiers.types.luis.LuisType",
-    "temba.classifiers.types.bothub.BothubType",
+    "temba.classifiers.types.bothub.BotHubType",
```
Comment (rowanseymour) [inline temba/settings_common.py]: I think you need to update your branch

File: .travis.yml (hunk 1 above)
Comment (hudsonbrendon) [reply .travis.yml]: I formatted the files that black accused error in travis, and even formatted, the error in travis persists, you know what it could be?

Comment (rowanseymour) [reply .travis.yml]: Maybe if you tell me what the error says 😄

Comment (hudsonbrendon) [reply .travis.yml]: development locally, it indicates these files to be formatted:

image

I formatted everyone using the command:

black --line-length=119 --target-version=py36 file

or 

black file ...

But when tests are run on travis, the error below happens:

image

could you tell what may be happening?

Comment (rowanseymour) [reply .travis.yml]: run..

```
black --line-length=119 --target-version=py36 temba
```

and then 

```
black --line-length=119 --target-version=py36 temba --check
```

to verify that there are no more changes to make

Comment (hudsonbrendon) [reply .travis.yml]: I executed the indicated commands:

image

however, the same errors persist in travis:

Screen Shot 2020-02-03 at 16 30 41

Comment (rowanseymour) [reply .travis.yml]: can you confirm you are using `black==19.3b0` as set in `pip-freeze.txt` ?

Comment (hudsonbrendon) [reply .travis.yml]: yes, it is exactly that version:

Screen Shot 2020-02-03 at 18 00 47

Comment (rowanseymour) [reply .travis.yml]: what does `black --version` output?

Comment (hudsonbrendon) [reply .travis.yml]: resolved, it was the black version, thank you very much!

**Comment by matmsa27** [comment]
Hi @rowanseymour , do you have any news about this PR?

**Comment by rowanseymour** [comment]
@matmsa27 you need to fix the missing coverage in unit tests and rebuild the locale files (those get out of date quickly and it's common for other PRs to touch those)

**Comment by matmsa27** [comment]
Hmmm, I got it @rowanseymour , we will fix it and commit the corrections. Thanks a lot.

File: temba/contacts/models.py [hunk 4]
```diff
@@ -38,8 +38,8 @@
 
 # phone number for every org's test contact
 OLD_TEST_CONTACT_TEL = "12065551212"
-START_TEST_CONTACT_PATH = 12_065_550_100
-END_TEST_CONTACT_PATH = 12_065_550_199
+START_TEST_CONTACT_PATH = "12_065_550_100"
+END_TEST_CONTACT_PATH = "12_065_550_199"
```
Comment (rowanseymour) [inline temba/contacts/models.py]: Not sure what this is for? These lines can actually be deleted

File: temba/contacts/models.py [hunk 5]
```diff
@@ -79,10 +80,11 @@
     (FCM_SCHEME, _("Firebase Cloud Messaging identifier"), FCM_SCHEME),
     (WHATSAPP_SCHEME, _("WhatsApp identifier"), WHATSAPP_SCHEME),
     (FRESHCHAT_SCHEME, _("Freshchat identifier"), FRESHCHAT_SCHEME),
+    (VK_SCHEME, _("VK identifier"), VK_SCHEME),
 )
 
 
-IMPORT_HEADERS = tuple((f"URN:{c[0]}", c[0]) for c in URN_SCHEME_CONFIG)
+IMPORT_HEADERS = tuple((f"URN:{c[0]}", c[0]) for c in URN_SCHEME_CONFIG)  # noqa
```
Comment (rowanseymour) [inline temba/contacts/models.py]: Is this necessary?

File: temba/contacts/models.py [hunk 6]
```diff
@@ -295,7 +301,7 @@ def normalize_number(cls, number, country_code):
         # now does it look plausible?
         try:
             if phonenumbers.is_possible_number(normalized):
-                return phonenumbers.format_number(normalized, phonenumbers.PhoneNumberFormat.E164), True
+                return (phonenumbers.format_number(normalized, phonenumbers.PhoneNumberFormat.E164), True)
```
Comment (rowanseymour) [inline temba/contacts/models.py]: Seems like your editor made some unrelated changes?

File: temba/contacts/models.py [hunk 7]
```diff
@@ -219,6 +221,10 @@ def validate(cls, urn, country_code=None):
                 except ValueError:
                     return False
 
+        # validate vk URNS look right (this is a guess)
+        elif scheme == VK_SCHEME:
```
Comment (rowanseymour) [inline temba/contacts/models.py]: if we don't know what the format of VK URNs is.. then maybe be more flexible here.. or don't do any validation

File: temba/contacts/models.py (hunk 5 above)
Comment (hudsonbrendon) [reply temba/contacts/models.py]: it wasn't, I fixed it.

File: temba/contacts/models.py (hunk 6 above)
Comment (hudsonbrendon) [reply temba/contacts/models.py]: probably yes, solved.

File: templates/channels/types/vk/claim.haml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+-block title
+  .medium-help.float-left{class:"{{ view.channel_type.icon }}"}
+
+  %h2.font_normalize.header-margin
+    {{ title }}
+
+-block pre-form
+  .info
+    %p
+      -blocktrans with name=brand.name
```
Comment (rowanseymour) [inline templates/channels/types/vk/claim.haml]: one last tiny change - I know you probably copied this from code that doesn't do this, but please add `trimmed` to your `blocktrans` tags to remove non-useful whitespace. 

note that if you merge with master, you should see a file called `code_check.py` in the project root which takes care of rebuilding locale etc.

Comment (hudsonbrendon) [reply templates/channels/types/vk/claim.haml]: @rowanseymour can you check my correction and merge if possible?

File: temba/channels/models.py [hunk 9]
```diff
@@ -261,7 +261,9 @@ class Channel(TembaModel):
     CONFIG_KEY = "key"
     CONFIG_API_ID = "api_id"
     CONFIG_API_KEY = "api_key"
+    CONFIG_CALLBACK_VERIFICATION_STRING = "callback_verification_string"
     CONFIG_CONTENT_TYPE = "content_type"
+    CONFIG_COMMUNITY_NAME = "community_name"
```
Comment (nicpottier) [inline temba/channels/models.py]: These should be on the type, not on the base models channel.

File: temba/channels/types/vk/type.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    code = "VK"
+    category = ChannelType.Category.SOCIAL_MEDIA
+
+    courier_url = r"^vk/(?P<uuid>[a-z0-9\-]+)/receive"
+
+    name = "VK"
+    icon = "icon-vk"
+
+    claim_blurb = _(
+        """Add a VK bot to send and receive messages on behalf of a VK community
```
Comment (nicpottier) [inline temba/channels/types/vk/type.py]: We usually have the name linked to the website, ie: https://vk.com/

File: temba/contacts/tests.py [hunk 11]
```diff
@@ -6883,22 +6883,24 @@ def test_json(self):
...
         ContactField.user_fields.filter(org=self.org, key="key0").update(label="AAAA")
 
         response = self.client.get(contact_field_json_url)
         response_json = response.json()
 
-        self.assertEqual(response_json[16]["label"], "AAAA")
-        self.assertEqual(response_json[16]["key"], "key0")
-        self.assertEqual(response_json[17]["label"], "First")
-        self.assertEqual(response_json[17]["key"], "first")
+        self.assertEqual(response_json[16]["label"], "Groups")
```
Comment (norkans7) [inline temba/contacts/tests.py]: For these tests I think you want to update the indexes in the list not the expected values

File: temba/channels/models.py (hunk 9 above)
Comment (hudsonbrendon) [reply temba/channels/models.py]: I use this in the channel's ClaimView form_valid method. In my opinion it is necessary. Note that in the model there is a field called "page_name", which is used for facebook, "community_name", it would be the same in vk.
```python
CONFIG_PAGE_NAME = "page_name"
```

```python
class ClaimView(ClaimViewMixin, SmartFormView):
    class Form(ClaimViewMixin.Form):
        community_access_token = forms.CharField(
            min_length=32, required=True, help_text=_("The Community Access Token")
        )
        community_name = forms.CharField(required=True, help_text=_("The name of the Community"))
        community_id = forms.IntegerField(required=True, help_text=_("The Community ID"))
        callback_verification_string = forms.CharField(required=True, help_text=_("The callback verification string"))

    form_class = Form

    def form_valid(self, form):
        org = self.request.user.get_org()
        community_access_token = form.cleaned_data["community_access_token"]
        community_name = form.cleaned_data["community_name"]
        community_id = form.cleaned_data["community_id"]
        callback_verification_string = form.cleaned_data["callback_verification_string"]

        config = {
            Channel.CONFIG_AUTH_TOKEN: community_access_token,
            Channel.CONFIG_COMMUNITY_NAME: community_name,
            Channel.CONFIG_SECRET: Channel.generate_secret(length=50),
            Channel.CONFIG_CALLBACK_VERIFICATION_STRING: callback_verification_string,
        }
        self.object = Channel.create(
            org, self.request.user, None, self.channel_type, name=community_name, address=community_id, config=config
        )

        return super().form_valid(form)
```

File: temba/channels/types/vk/type.py (hunk 10 above)
Comment (hudsonbrendon) [reply temba/channels/types/vk/type.py]: Resolved.

File: temba/contacts/tests.py (hunk 11 above)
Comment (hudsonbrendon) [reply temba/contacts/tests.py]: It was the indices, I added a value in the URN_SCHEME_CONFIG present in the models of the contact app and I needed to update the values.

**Comment by matmsa27** [comment]
Hi @nicpottier , any news about this PR?

File: temba/channels/models.py (hunk 9 above)
Comment (nicpottier) [reply temba/channels/models.py]: Not sure I understand, the comment I have is that config keys that are specific to a channel subtype do not belong on the global Channel type, but instead should live on the subtype.

IE: see https://github.com/nyaruka/rapidpro/blob/master/temba/channels/types/whatsapp/type.py#L91

`CONFIG_COMMUNITY_NAME` being used in the Facebook channel is a bug, not a feature, that should also be moved to the channel subtype. Just move these over to the type so we can keep this namespace clean.

**Comment by nicpottier** [comment]
Resolve conflicts, move the CONFIG params and we are good to go. Please do that today if at all possible so we can merge for 5.4 inclusion.

**Comment by nicpottier** [comment]
Thanks!

**Comment by matmsa27** [comment]
Thanks a lot @nicpottier :)
